R/utils.R

```diff
@@ -167,34 +167,70 @@ repair_variable_names <- function(names) {
   names
 }
 
-
-
-
 #' Dump data to temporary file in format readable by CmdStan
 #'
-#' Currently calls `rstan::stan_rdump()` to create the `.data.R` file.
-#' FIXME:
-#'
 #' @param data A named list of \R objects.
 #' @return Path to temporary file containing the data.
 #' @noRd
-write_rdump <- function(data) {
-  # FIXME don't use rstan
-  if (!requireNamespace("rstan", quietly = TRUE)) {
-    stop("Please install the 'rstan' package. This is temporarily required ",
-         "for writing the data file for CmdStan.", call. = FALSE)
+write_json <- function(data) {
+  if (!requireNamespace("jsonlite", quietly = TRUE)) {
+    stop("Please install the 'jsonlite' package.", call. = FALSE)
   }
   checkmate::assert_names(names(data), type = "unique")
-  temp_file <- tempfile(pattern = "standata-", fileext = ".data.R")
-  rstan::stan_rdump(
+  temp_file <- tempfile(pattern = "standata-", fileext = ".json")
+  cmdstanr_jsondump(
     list = names(data),
     file = temp_file,
     envir = as.environment(data)
   )
   temp_file
 }
 
-
+#' Dump data to temporary file in format readable by CmdStan
+#'
+#' @param list A vector of character string: the names of one or more R objects to be dumped.
+#' @param file A character string represeneting the path to the output file
+#' @param envir The environment to search for the listed objects.
+#' @export
+#' @return Path to temporary file containing the data.
+cmdstanr_jsondump <- function(list,
+                          file,
+                          envir = parent.frame()) {
+  if (!requireNamespace("jsonlite", quietly = TRUE)) {
+    stop("Please install the 'jsonlite' package.", call. = FALSE)
+  }
+  temp_file <- tempfile(pattern = "standata-", fileext = ".json")
+  # check filename is valid(vector of characters) and of nonzero length
+  if (!is.character(file) || !nzchar(file)) {
+    warning("The supplied filename is invalid!")
+    return(invisible(character()))
+  }
+  # check if all variables in list exist
+  variable_exists <- sapply(list, exists, envir = envir)
+  if(!all(variable_exists)) {
+    warning(paste("The following objects were not found: ", paste(list[!variable_exists], collapse = ", ")))
+    return(invisible(character()))
+  }
+  data = list()
+  for (var_name in list) {
+    var <- get(var_name, envir)
+    if(!(is.numeric(var) || is.factor(var) || is.logical(var) || is.data.frame(var)  || is.list(var))) {
+      warning(paste("Variable ", var_name, " is of invalid type."))
+      return(invisible(character()))
+    }
+    # convert TRUE/FALSE to 1/0
+    if(is.logical(var)) {
+      mode(var) <- "integer"
+    } else if(is.data.frame(var)) {
+      var <- data.matrix(var)
+    }
+    data[[var_name]] <- var
+  }
+  # call to write JSON with
+  # unboxing variables (N = 10 is stored as N : 10, not N: [10])
+  # handling factors as integers
+  jsonlite::write_json(data, auto_unbox = TRUE, factor = "integer", path = file)
```

Comment:
`jsonlite::write_data()` returns `NULL` if you assign it to a value like 

```
x <- jsonlite::write_json(...) # x is NULL
```

which is fine, but currently the doc for `cmdstanr_jsondump` says it returns a file path (looks just like a copy and paste issue). 

Assuming `cmdstan_jsondump()` is exposed to the user I don't really know if it makes sense to return anything to the user from this function or not. `stan_rdump()` just returns the names of the objects it wrote I think, but I don't know if that's super useful. And we don't need it to return a file path since it's just the value of the `file` argument. So I guess `NULL` is fine, just change the doc? What do you think?

Reply:
Yes, a copy paste issue. Will fix.